compressors/arithmetic_coding.py

```diff
@@ -435,3 +511,96 @@ def test_adaptive_arithmetic_coding():
         lossless_entropy_coder_test(
             encoder, decoder, freq, DATA_SIZE, encoding_optimality_precision=1e-1, seed=0
         )
+
+
+def _generate_2nd_order_markov(num_samples: int, seed: int = 0):
+    """generate a 2nd order Markov distribution for testing.
+
+    Defined on alphabet {0,1,2}, the distribution is defined like
+    X_n = X_{n-1} + X_{n-2} + Ber(1/2) mod 3
+
+    The entropy rate is 1 bit/symbol.
+
+    The stationary distribution is the uniform distribution.
+    """
+    assert num_samples >= 3
+    rng = np.random.default_rng(seed)
+    random_bits = rng.choice(2, size=num_samples - 2)
+    markov_samples = np.zeros(num_samples, dtype=np.uint8)
+    markov_samples[0] = rng.choice(3)
+    markov_samples[1] = rng.choice(3)
+    for i in range(2, num_samples):
+        markov_samples[i] = (markov_samples[i - 1] + markov_samples[i - 2] + random_bits[i - 2]) % 3
+    return DataBlock(markov_samples)
+
+
+def _test_adaptive_order_k(
+    encoder: DataEncoder,
+    decoder: DataDecoder,
+    data_block: DataBlock,
+    expected_bitrate: float,
+    encoding_optimality_precision: float,
+):
+    """Checks if the adaptive order k arithmetic coding performs lossless compression and if the performance is
+    close to the expectation.
+
+    Args:
+        encoder (DataEncoder): Encoder to test with
+        decoder (DataDecoder): Decoder to test lossless compression with
+        data_block (DataBlock): data to use for testing
+        expected_bitrate (float): the theoretically expected bitrate
+        encoding_optimality_precision (float): check that the average expected_bitrate is close to the avg_codelen
+    """
+    # check if encoding/decoding is lossless
+    is_lossless, encode_len, _ = try_lossless_compression(
+        data_block, encoder, decoder, add_extra_bits_to_encoder_output=True
+    )
+
+    # avg codelen ignoring the bits used to signal num data elements
+    avg_codelen = (encode_len) / data_block.size
+    print(f" expected_bitrate={expected_bitrate:.3f}, avg_codelen: {avg_codelen:.3f}")
+
+    # check whether arithmetic coding results are close to expected codelen
+    err_msg = f"avg_codelen={avg_codelen} is not {encoding_optimality_precision} close to expected_bitrate={expected_bitrate}"
+    assert np.abs(avg_codelen - expected_bitrate) < encoding_optimality_precision, err_msg
+
+    assert is_lossless
+
+
+def test_adaptive_order_k_arithmetic_coding():
+    """
+    Test AEC coding on 2nd order Markov
+    - Check if encoding/decodng is lossless
+    - Check if the compression is close to expected for k = 0, 1, 2, 3
+    - Verify that 0th order matches the adaptive IID exactly.
+    """
+
+    DATA_SIZE = 10000
+
+    data_block = _generate_2nd_order_markov(DATA_SIZE)
+
+    # for the source as defined in _generate_2nd_order_markov, the
+    # 0th and 1st order entropy is log_2(3) [uniform] and beyond that it is 1.
+
+    print()
```

Comment:
typo?

Reply:
no, it was to properly format the test output. Added a comment now.